### ISSUE TITLE
Implementation of ReactComponentEnvironment.replaceNodeWithMarkupByID

### ISSUE THREAD

**Issue by yuchi** [Yomguithereal/react-blessed#5]
You know that because of `react-titanium` we share a lot of common pains :)

Well, looks like that:

1. In `react@0.14.0-beta3` we still cannot re-inject *ReactComponentEnvironment*.
2. In some specific cases we need to implement `replaceNodeWithMarkupByID` otherwise it will throw errors at us.

A simple test case that worked (which means failed!) for me is:

```js
class Simple extends Component { /* ... */ }
class Advanced extends Component { /* ... */ }

class Root extends Component {
  render() {
    return this.state.something ? <Simple /> : <Advanced />;
  }
}
```

and then change the value of `state.something` to a truthy and React will understand that it must completely replace the nodes rendered by `Advanced` and calls `replaceNodeWithMarkupByID` which of course doesn’t work in our scenarios.

May I suggest you to check if you need to implement it too?

**Comment by Yomguithereal** [comment]
Hello @yuchi, let me check this one thoroughly. If I remember correctly, even if you cannot inject *ReactComponentEnvironment` you can still override its properties/methods as a dirty workaround.

**Comment by yuchi** [comment]
That’s what I did on yuchi/react-titanium@b1da1f26d94d160fbc1adcbe9c6533c5b007e2f4 but I’m not exactly sure about what should I do with it.

In my case I was replacing a top level window so a no-op is ok for me, but I can’t foresee when this will bite me again.

**Comment by sidorares** [comment]
Hey @yuchi @Yomguithereal - could you recommend a good starting point for implementing custom react renderer? I've looked briefly at react-canvas, react-three and react-hardware sources and all have slightly different approach. I'd like to create "react-x11" renderer on top of my [x11](https://github.com/sidorares/node-x11) X Window client. As far as I understand this is not documented anywhere and the only source of information is react internals and existing renderers? Have you came across any documentation/posts on that topic?

**Comment by yuchi** [comment]
I was tinkering with the idea of writing a guide. For me reading the source of both react-dom client and react-blessed was enough.

**Comment by yuchi** [comment]
But there are parts that still don't grasp completely, the pooled transactions for example. L

**Comment by sidorares** [comment]
thanks! Please do the guide!

Here is my list of custom integrations:

https://github.com/Yomguithereal/react-blessed
https://github.com/Flipboard/react-canvas
https://github.com/iamdustan/react-hardware
https://github.com/Izzimach/react-pixi
https://github.com/Izzimach/react-three/
https://github.com/yuchi/react-titanium
https://github.com/reactjs/react-art
https://github.com/facebook/react-native

at the moment react-blessed and react-hardware looks like the best starting points

**Comment by Yomguithereal** [comment]
Hello @sidorares. If I remember correctly there are several approach to making a custom React renderer. Quite like pre-0.14 and 0.14. Now that the rendering routine has been separated from the core, it has been easier to develop custom rendering routine. So you must have, on one side, custom renderers using hacks and or creating custom components using no DOM elements while being rendered by the DOM renderer anyway. And, on the other side, renderers like `react-titanium` and `react-blessed` using 0.14's logic. The downside is of course that you cannot use a pre 0.14 version with it.

**Comment by Yomguithereal** [comment]
Anyway, for the time being, and until @yuchi creates a guide (for which I'll gladly provide feedback and help), don't hesitate to come back here for precisions and so we can tackle incoming problems together.

**Comment by gaearon** [comment]
I'm getting these too:

```js
TypeError: Cannot read property 'firstChild' of undefined
    at Object.ReactMount.findComponentRoot (/Users/dan/p/react-blessed-hot-motion/node_modules/react/lib/ReactMount.js:640:39)
    at Object.ReactMount.findReactNodeByID (/Users/dan/p/react-blessed-hot-motion/node_modules/react/lib/ReactMount.js:586:23)
    at Object.getNode (/Users/dan/p/react-blessed-hot-motion/node_modules/react/lib/ReactMount.js:160:32)
    at Object.ReactDOMIDOperations.dangerouslyReplaceNodeWithMarkupByID (/Users/dan/p/react-blessed-hot-motion/node_modules/react/lib/ReactDOMIDOperations.js:125:27)
    at Object.wrapper [as replaceNodeWithMarkupByID] (/Users/dan/p/react-blessed-hot-motion/node_modules/react/lib/ReactPerf.js:66:21)
    at [object Object].ReactCompositeComponentMixin._replaceNodeWithMarkupByID (/Users/dan/p/react-blessed-hot-motion/node_modules/react/lib/ReactCompositeComponent.js:538:31)
    at [object Object].ReactCompositeComponentMixin._updateRenderedComponent (/Users/dan/p/react-blessed-hot-motion/node_modules/react/lib/ReactCompositeComponent.js:530:12)
    at [object Object].ReactCompositeComponentMixin._performComponentUpdate (/Users/dan/p/react-blessed-hot-motion/node_modules/react/lib/ReactCompositeComponent.js:503:10)
    at [object Object].ReactCompositeComponentMixin.updateComponent (/Users/dan/p/react-blessed-hot-motion/node_modules/react/lib/ReactCompositeComponent.js:432:12)
    at [object Object].wrapper [as updateComponent] (/Users/dan/p/react-blessed-hot-motion/node_modules/react/lib/ReactPerf.js:66:21)
```

Why does this happen? Is there a hook missing on React side, or is there something this implementation is currently missing?

**Comment by yuchi** [comment]
That's exactly the kind of errors that pushed me into implementing it as a noop.

**Comment by Yomguithereal** [comment]
It is currently a noop too in `react-blessed` but it seems to fail anyway. I need to investigate this further. @gaearon can you provide us with the code that led to this error please?

**Comment by gaearon** [comment]
Sure. This is enough:

```js
export default class App extends Component {
  componentDidMount() {
    setInterval(this.forceUpdate.bind(this), 200);
  }

  render() {
    if (Math.random() > 0.5) {
      return <box key='a'>lol</box>;
    } else {
      return <box key='b'>oops</box>;
    }
}
```

Note how keys are different, so reconciliation is forced.

**Comment by gaearon** [comment]
I seem to be able to fix it by adding [after this line](https://github.com/Yomguithereal/react-blessed/blob/7e59e81f80d9afc45b62eaf264fca3626236aa2e/src/ReactBlessedInjection.js#L32):

```js
// NOTE: very dirty trick due to react@0.14-beta1's current state
ReactComponentEnvironment.processChildrenUpdates = Function.prototype;
```

one more line:

```js
ReactComponentEnvironment.replaceNodeWithMarkupByID = Function.prototype;
```

**Comment by gaearon** [comment]
So the problem is `ReactInjection.Component.injectEnvironment()` can be called only once. It's already being called from React entry point because `react` and `react-dom` are still a single thing inside since they're still shipped together for backward compat until 0.15. Therefore we can't call it again, and have to resort to monkeypatching `ReactComponentEnvironment` directly.

Am I correct? @spicyj is there a plan to solve this before 0.14?

**Comment by spicyj** [comment]
I think that sounds right. There is not.

**Comment by gaearon** [comment]
Thanks for confirming!

**Comment by gaearon** [comment]
I sent https://github.com/Yomguithereal/react-blessed/pull/16 as temporary fix until 0.15.

**Comment by Yomguithereal** [comment]
Will check the PR now @gaearon. Thanks.
